diff --git a/swig/swig_interfaces/oc_api.i b/swig/swig_interfaces/oc_api.i
index e600456c..47ff85cb 100644
--- a/swig/swig_interfaces/oc_api.i
+++ b/swig/swig_interfaces/oc_api.i
@@ -1372,7 +1372,7 @@ bool jni_oc_init_post(const char *uri, oc_endpoint_t *endpoint, const char *quer
     OC_DBG("JNI - initPost failed releasing jni_sync_lock");
     OC_DBG("JNI: - unlock %s\n", __func__);
   }
-  return returnValue
+  return returnValue;
 }
 %}
 
